Add the `Origin-Isolation` header

domenic opened the pull request:
See https://github.com/WICG/origin-isolation.

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chromium
   * [Gecko](https://mozilla.github.io/standards-positions/#domenic-origin-isolation)
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/tree/master/origin-isolation
   * More coming: https://github.com/WICG/origin-isolation/issues/27
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1042415
   * Firefox: …
   * Safari: …

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

---

Things that I need help on:

- This has a similar problem as #5334 in that it wants to do a secure context check before the environment settings object/Window/Document are created. It seems like we'll need to refactor the secure contexts spec a good bit, for both of them... But I'd like someone to confirm that line of thinking before I go off and do so. /cc @mikewest

- I'm not sure if changing the agent cluster key automatically makes `document.domain` a no-op, or not. I think by my current reading, [IsPlatformObjectSameOrigin](https://html.spec.whatwg.org/#isplatformobjectsameorigin-(-o-)) does not care about agent clusters, so this PR as-written does not prevent cross-agent-cluster sync access. Should we add an agent cluster check to IsPlatformObjectSameOrigin, and let the domain component change via the `document.domain` setter? Or should we store an "is origin-isolated" boolean on the Document, and bail out of the `document.domain` setter algorithm if that boolean is true? The former seems a bit nicer, but I'm unsure if there's any observable difference between the two. /cc @dtapuska

domenic commented:
Hmm, I think there is a big difference between just changing IsPlatformObjectSameOrigin, and storing a boolean on the document. If we only change IsPlatformObjectSameOrigin, then other same origin-domain checks will still pass. So I think we need to store a boolean, and also write tests for other such checks.

annevk commented:
I like the idea though that changing the keying of the agent cluster is what disables document.domain. Unfortunately there's no clear way back from an agent cluster to its key. Setting a flag on the agent cluster similar to cross-origin isolated is something you could do though. That also avoids having to maintain a per-document boolean.

domenic commented:
Did you see my latest commit? I think with the historical agent cluster key map, there kind of effectively is such a way back now.

annevk commented on source:
```diff
@@ -81993,10 +82059,20 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+     data-x="concept-response-header-list">header list</span>.</p></li>
+
+     <li><p>Let <var>requestsOI</var> be true if <var>oiHeader</var> is not failure,
+     <var>oiHeader</var> is not null, and <var>oiHeader</var>[0] is the <span>structured header
```
Why is _oiHeader_ indexable as such?

domenic replied:
All structured header items are a pair of the actual value plus the parameters.

annevk replied:
I guess Fetch should define how they translate to Infra then if we want to do this. Might also be nice for parameters.

domenic replied:
I think the concepts used are already Infra concepts: pairs, maps, booleans, ...

annevk replied:
Heh, if only they would have referenced it as well. But okay, that should make any formal equating pretty straightforward.

annevk replied:
I noticed you still used "structured header boolean true". If they are the same we don't have to do that, right?

annevk commented on source:
```diff
@@ -79586,6 +79625,62 @@ interface <dfn>BarProp</dfn> {
+  <p>Note that the `<code data-x="http-origin-isolation">Origin-Isolation</code>` header can never
+  cause same-origin <code>Document</code> objects to end up in different <span data-x="agent
+  cluster">agent clusters</span>, even if one sends the header and the other doesn't. This is
+  prevented by means of the <span>historical agent cluster key map</span>.</p>
```
This needs to be preceded by "Within a browsing context group, the ..."

domenic replied:
Happy to add it, although it doesn't seem necessary since agent clusters don't span BCGs.

annevk replied:
I might be missing something, but my point is that same-origin Document objects can end up in different agent clusters.

domenic replied:
Hmm yeah, I see what you mean. OK cool. Pushed now.

annevk commented on source:
```diff
@@ -79586,6 +79625,62 @@ interface <dfn>BarProp</dfn> {
+
+   <dd>
+    <p>Returns true if the origin isolation restrictions are in place for this <code>Window</code>,
+    i.e. if the `<code data-x="http-origin-isolation">Origin-Isolation</code>` header has been used
```
i.e.,

annevk commented on source:
```diff
@@ -82356,10 +82451,20 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+
+     <li><p>Let <var>requestsOI</var> be true if <var>oiHeader</var> is not failure,
+     <var>oiHeader</var> is not null, and <var>oiHeader</var>[0] is the <span>structured header
+     boolean</span> true.</p></li>
```
It's too early, but I can't help but think we want an abstraction to make this less awkward and encourage consistency in handling. No great ideas other than collapsing failure into null though. (Or maybe that is what we should always do?)

domenic replied:
I do suspect collapsing failure into null might be what all callers want. Perhaps let's reassess after this lands, COOP lands, COEP lands, and CORP gets converted to a structured header. (And maybe by then Permission Policy / Document Policy might be ready too.)

annevk commented on source:
```diff
@@ -88739,41 +88844,78 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
   of <span data-x="similar-origin window agent">similar-origin window agents</span>.</p>
 
-  <p>An <dfn>agent cluster key</dfn> is a <span>site</span>.</p>
+  <p>An <dfn>agent cluster key</dfn> is a <span>site</span> or an <span>origin</span>.</p>
```
I prefer the wording https://github.com/whatwg/html/pull/4734 has for this as it's more precise.

domenic replied:
I found that wording quite confusing (if it can be any origin, just state that, like I did here?) but I'll copy it over anyway.

domenic replied:
I tweaked it a bit since the previous version seemed to imply that keys could not be site-and-registrable-domains.

annevk commented on source:
```diff
@@ -88739,41 +88844,78 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
+
+  <p class="XXX"><a href="https://github.com/whatwg/html/pull/4734">Issue #4734</a> is expected to
+  further modify this algorithm to increase the number of cases where the returned value is an
+  <span>origin</span>.</p>
```
I prefer the approach in that PR by modifying "obtain a similar-origin window agent" and inlining "obtain an agent cluster key".

I also wonder, if we did that, could we modify browsing context group's historical agent cluster key map before invoking that algorithm so it could use that map to make decisions, requiring us to pass less state around?

Furthermore, now that we tie state to browsing context groups, it almost seems like we should have different types of them. Those that carry a historical map and those that are cross-origin isolated. But some asserts is probably enough initially.

domenic replied:
https://github.com/whatwg/html/pull/4734/files#r405428958 indicates we want to fold all three algorithms together. I'll give that a shot...

domenic replied:
This turned out pretty nicely.

I didn't address your latest paragraph so I'll leave this un-resolved. If you have concrete suggestions (either for now or as a followup) then let me know; otherwise I wasn't sure where to go with it.

annevk replied:
I was mostly wondering if we should make normal browsing context groups and cross-origin isolated browsing context groups more clearly mutually exclusive. We can address it in the context of cross-origin isolated.

annevk commented:
It's unfortunate PR Preview no longer seems to work here. I added a space to try to trigger it, but it's not working.

annevk commented on source:
```diff
@@ -88739,58 +88845,67 @@ import "https://example.com/foo/../module2.mjs";</code></pre>
-  <p>To <dfn data-x="obtain-browsing-agent-cluster">obtain a browsing context agent cluster</dfn>,
-  given a <span>browsing context group</span> <var>group</var> and <span>agent cluster key</span>
-  <var>key</var>, run these steps:</p>
+    <p>then set <var>key</var> to <var>origin</var> and set <var>oiRestricted</var> to true.</p>
```
You also need to write into the map here.

(My other variant of writing into the map before you obtain an agent has the nice property of not needing _requestsOI_, but you end up writing into the map even when you don't need to (i.e., when site is equal to origin), so this is probably the way to go.)

domenic commented:
> My initial take was that cross-origin isolated implies the latter, but this would do the wrong thing for things that are already keyed on origin. Perhaps we should revisit that as allowing those to be origin-isolation restricted despite it not meaning anything would simplify several things at this point?

The way I was envisioning it was that cross-origin isolated just causes _requestsOI_ to be true (regardless of the `Origin-Isolation` header). Then the logic in this PR takes over and _oiRestricted_ (and thus `window.originIsolationRestricted`) will be false if you were already keyed to an origin. But `window.crossOriginIsolated` would be true.

WDYT?

annevk commented:
That could work I think. We'd change

> ` <li><var>requestsOI</var> is true</li>`

to also check for bcg's cross-origin isolated and where we write into the map (TBD) we add another conditional to only do so if bcg's cross-origin isolated is false.

domenic commented:
> where we write into the map (TBD) we add another conditional to only do so if bcg's cross-origin isolated is false.

I don't think this would be strictly necessary, because cross-origin isolation ensures consistency within a BCG itself. So the map would just be redundant for any BCG with cross-origin isolated pages; writing to it would be fine (and have no effect).

I guess maybe that gets to your point above about making normal BCGs and COI BCGs mutually exclusive.

annevk commented:
If we did that I think we'd have to add a note, but maybe we should do that either way. (Agreed that there is no observable difference.)

domenic commented:
This now has a secure context check. It, like COOP, needed to check the response's HTTPS state, so maybe we should have an abstraction? There's no great one though, since _reservedEnvironment_ and _response_ are separate anyway. The best we could do is something like

> If _request_ and _reservedEnvironment_ do not **represent a new secure context**, then...

although I think there should be better names.

Anyway, this is now blocked on https://github.com/WICG/origin-isolation/issues/31, I guess.

domenic commented:
I updated this on top of #4734, and also put it on top of #5714 (although that's not essential). I'd love another round of review.

In particular I changed the spec per https://github.com/WICG/origin-isolation/issues/31#issuecomment-649745305 (and will go change the explainer now).

The changes to "obtain a similar-origin window agent" turned out a bit different than I expected, and there might be a slightly less awkward way to phrase them, but I think they work.

annevk commented on source:
```diff
@@ -79840,6 +79866,66 @@ interface <dfn>BarProp</dfn> {
+   </dd>
+  </dl>
+
+  <p>A <code>Document</code> delivered over a <span>secure context</span> can opt in to origin
```
into? We only have two instances of "in to" and both look suspect.

domenic replied:
I think for cases like "zoom in" or "opt in" it's best to keep those phrases separate from the "to". There's nothing very concrete on this online; e.g. https://english.stackexchange.com/questions/422848/opt-into-vs-opt-in-to is unanswered and https://forum.wordreference.com/threads/opt-in-or-opt-into.3560327/ is wishy washy. The headline at https://www.wsj.com/articles/more-parents-opt-into-opt-out-1429893574 separates them at least.

annevk commented on source:
```diff
@@ -79840,6 +79866,66 @@ interface <dfn>BarProp</dfn> {
+  prevented by means of the <span>historical agent cluster key map</span>.</span></p>
+
+  <p class="note">This means that the <code data-x="dom-originIsolated">originIsolated</code> getter
+  could return false, even if the header is set, if the header was omitted on a previously-loaded
```
can instead of could? (both times)

annevk commented on source:
```diff
@@ -87139,8 +87273,22 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+     <li>
+      <p>Set <var>agentCluster</var>'s <span>origin-isolated</span> to true.</p>
+
+      <p class="note">These workers/worklets can be considered to be origin-isolated. However, this
```
For a worklet isTopLevel is currently always false, although I'm not sure that should be the architecture for non-audio worklets. It does make this note a bit confusing however.

domenic replied:
Good catch; I'll change it to just "workers" for now.

annevk commented on source:
```diff
@@ -87090,16 +87193,44 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+     <li><p>Set <var>group</var>'s <span>historical agent cluster key map</span>[<var>origin</var>]
+     to <var>key</var>.</p></li>
+    </ol>
+   </li>
```
If we change this whole addition to start with "Otherwise, if" we could keep the historical agent cluster key map and cross-origin isolated separated. I would somewhat prefer that as the less (unneeded) bookkeeping the better.

(It would also allow more easily for introducing different group types in the future.)

domenic replied:
Hmm, I'll give it a try.

annevk commented on source:
```diff
@@ -87090,16 +87193,33 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+   <var>group</var>'s <span>historical agent cluster key map</span>[<var>origin</var>].</p></li>
+
+   <li>
+    <p>Otherwise, if <var>requestsOI</var> is true, then:</p>
```
This part of the change is wrong. I initially read the algorithm like this, but that ends up making the map redundant as it would only ever contain origin values. The _requestsOI_ conditional has to be part of overwriting the key only.

domenic replied:
I think I misunderstood what you meant by keeping them separate; PTAL.

annevk replied:
This works, but I think what would also work is keeping this as

Otherwise:

* If requestsOI is true, set ...
* Modify map.

My "Otherwise, if" comment was about the step above this one.

domenic replied:
Ah, thanks, that makes sense now. Done.

domenic commented:
Alright, I've updated all the tests to use the standard `window.originIsolated` getter. And, the test suite is basically finished. (A couple PRs around popups are still going through review but they should land soon.) So, I'm now comfortable with this getting merged.

@annevk, want to do the final review?
